File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        st = []
        res = 0
        for i in range(len(heights)):
            index = i
            while st and heights[i] <= st[-1][0]:
                area = st[-1][0] * ( i - st[-1][1])
                res = max(res,area)
                index = st[-1][1]
                st.pop() 
            st.append((heights[i], index))
        while st: 
            area2 = st[-1][0] * ( len(heights) - st[-1][1])
            res = max(area2,res)
            st.pop()
        return res

File: Data Structures & Algorithms/spiral-matrix/submission-1.cpp
class Solution {
public:
    
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
            vector<int> res;
    int a = 0, b = matrix[0].size()-1;
    int c = matrix.size()-1, d = 0;
    
    while(a <= c && d <= b){
        
        for(int i = d; i <= b; i++){ 
            res.push_back(matrix[a][i]);
        }
        
        for(int i = a+1; i <= c; i++) {
            res.push_back(matrix[i][b]);
        }
        
        
        if(a < c)
            for(int i = b-1; i >= d; i--){ 
                res.push_back(matrix[c][i]);
            }
        
        if(d < b)
            for(int i = c-1; i > a; i--){ 
                res.push_back(matrix[i][d]);
            }
        a++; 
        b--; 
        c--; 
        d++;
    }
    return res;
    }
};
